### PR TITLE
Add APIResourceSchema and APIExport KCP objects for managed-gitops (Solves: #GITOPSRVC-188) 

### PR DESCRIPTION
**Pull request by samyak-jn**
#### Description:
- Initial support for KCP by adding objects APIResourceSchema for all our CRDs
- Addition of APIExport linking all APIResourceSchemas
- Thanks to the `has team` for providing the starter script `generate-kcp-api.sh` for getting started, tweaks in the script to work with our repository involve path correction since we are on our monorepo our CRDs lie in `backend-shared/config/*` and `appstudio-shared/config/*`. The kcp objects are created in the same path `backend-shared/config/kcp/`
- Made a change to the script to point to managed-gitops instead of appstudio as per our CRDs support (`s/appstudio/managed-gitops/g`) for backend-shared component
- NOTE: A pre-requisite for this script is to have kubectl-kcp installed

### CODE REVIEW

File: utilities/generate-kcp-api.sh [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+if ! which kubectl-kcp; then
+  echo "kubectl-kcp required on path"
+  echo "you can install it running:"
+  echo "    $ git clone https://github.com/kcp-dev/kcp && cd kcp && make install"
+  exit 1
+fi
+
+THIS_DIR="$(dirname "$(realpath "$0")")"
+CRD_DIR="$(realpath ${THIS_DIR}/../backend-shared/config/crd/bases)"
```
Comment (chetan-rns) [inline utilities/generate-kcp-api.sh:13]: I suppose we should generate these resources for other appstudio CRDs as well. `appstudio-shared/config/crd`

File: backend-shared/config/kcp/apiresourceschema_gitopsrvc.yaml [hunk 2]
```diff
@@ -0,0 +1,436 @@
+# This file is generated from CRDs by ./hack/generate-kcp-api.sh script.
+# Please do not modify!
```
Comment (chetan-rns) [inline backend-shared/config/kcp/apiresourceschema_gitopsrvc.yaml:2]: Please include a Makefile target to regenerate these files when the CRD changes.

File: backend-shared/config/kcp/apiexport_gitopsrvc.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# This file is generated from CRDs by ./hack/generate-kcp-api.sh script.
+# Please do not modify!
+
+apiVersion: apis.kcp.dev/v1alpha1
+kind: APIExport
+metadata:
+  name: gitopsrvc
+spec:
+  latestResourceSchemas:
```
Comment (chetan-rns) [inline backend-shared/config/kcp/apiexport_gitopsrvc.yaml:9]: Do we need to include [permissionClaims](https://github.com/kcp-dev/kcp/blob/6d2008e2fdeebbb3ef7874e7d1a559fca4e1ac53/pkg/apis/apis/v1alpha1/types_apiexport.go#L157) for secrets and other resources? I'm hitting this error when testing it with controllers


```shell
2022-07-28T06:05:23.610Z    ERROR    controller-runtime.source    source/source.go:137    if kind is a CRD, it should be installed before calling Start    {"kind": "Secret", "error": "no matches for kind \
"Secret\" in version \"v1\""}
```

Comment (samyak-jn) [reply backend-shared/config/kcp/apiexport_gitopsrvc.yaml]: As per the findings, it is agreed to make the gitops-service controller happy we need at least namespaces secrets to get this working, but to address this another problem I'm referencing this issue #198 here as well (separate from the 188 stories)

**Comment by chetan-rns** [comment]
@samyak-jn  Looks like we need APIResourceSchemas for Appstudio Application and Argo CD Application CRDs. We do not maintain these CRDs in our repository instead we apply them directly  

https://github.com/redhat-appstudio/managed-gitops/blob/849975dad51b45e6347e541c891d6751d951461b/Makefile#L75

https://github.com/redhat-appstudio/managed-gitops/blob/849975dad51b45e6347e541c891d6751d951461b/Makefile#L108

Error:
```
14:12:57        cluster-agent | 2022-08-11T14:12:57.699985356+05:30	ERROR	controller-runtime.source	if kind is a CRD, it should be installed before calling Start	{"kind": "Application.argoproj.io", "error": "no matches for kind \"Application\" in version \"argoproj.io/v1alpha1\""}

14:12:33 appstudio-controller | 2022-08-11T14:12:33.506977721+05:30	ERROR	controller-runtime.source	if kind is a CRD, it should be installed before calling Start	{"kind": "Application.appstudio.redhat.com", "error": "no matches for kind \"Application\" in version \"appstudio.redhat.com/v1alpha1\""}
```

**Comment by samyak-jn** [comment]
Hi @chetan-rns 

I added APIResourceSchema and APIExport object for the same as mentioned above in dependencies dir under  `$(ROOT_DIR)/mainfests`.
As per the conversation, instead of creating a customization file, I created a make target `make apply-kcp-api-all` which applies all the APIExport files in one go, LMK if this works for you or resolves the problem at upper hand.

File: Makefile [hunk 4]
```diff
@@ -203,6 +203,13 @@ setup-e2e-local: install-argocd-openshift devenv-docker reset-db ## Setup steps
 
 start-e2e: start ## Start the managed gitops processes for E2E tests. At the moment this is just a wrapper over 'start' target
 
+gen-kcp-api-all: gen-kcp-api-appstudio-shared gen-kcp-api-backend-shared ## Creates all the KCP API Resources for all comfig/crds
+
+apply-kcp-api-all: ## Apply all APIExport to the cluster
+	kubectl apply -f manifests/dependencies/kcp/apiexport_gitopsrvc_dependencies.yaml
+	kubectl apply -f appstudio-shared/config/kcp/apiexport_gitopsrvc_appstudioshared.yaml
+	kubectl apply -f backend-shared/config/kcp/apiexport_gitopsrvc_backendshared.yaml
+
```
[the remark this answers is not in the document]
Comment (chetan-rns) [reply Makefile:212]:
```shell
[cbanavik@cbanavik managed-gitops]$ make apply-kcp-api-all
kubectl apply -f manifests/dependencies/kcp/apiexport_gitopsrvc_dependencies.yaml
apiexport.apis.kcp.dev/gitopsrvc-dependencies unchanged
kubectl apply -f appstudio-shared/config/kcp/apiexport_gitopsrvc_appstudioshared.yaml
apiexport.apis.kcp.dev/gitopsrvc-appstudio-shared unchanged
kubectl apply -f backend-shared/config/kcp/apiexport_gitopsrvc_backendshared.yaml
apiexport.apis.kcp.dev/gitopsrvc-backend-shared unchanged

[cbanavik@cbanavik managed-gitops]$ kubectl get apiresourceschemas.apis.kcp.dev
No resources found
```

I think we overlooked the above comment Samyak. Can we apply APIResourceSchema resources?

[the remark this answers is not in the document]
Comment (samyak-jn) [reply Makefile]: Hey @chetan-rns ,
Sorry, I forgot to add the changes from the makefile to the latest commit, fixed now.

**Comment by chetan-rns** [comment]
Thanks, @samyak-jn. I tried the latest changes with the controllers and I don't see any errors. We could also get rid of ApiExport in the dependencies folder since it won't be used.

/lgtm

**Pull request merged**